File: algorithms/21_scala/dp/Kadane.scala
object Kadane {
  def kadane(arr: Array[Int]): Int = {
    var best = arr(0)
    var cur = arr(0)
    for (i <- 1 until arr.length) {
      cur = math.max(arr(i), cur + arr(i))
      best = math.max(best, cur)
    }
    best
  }

  def main(args: Array[String]): Unit = {
    assert(kadane(Array(-2, 1, -3, 4, -1, 2, 1, -5, 4)) == 6, "kadane wrong")
    assert(kadane(Array(-5, -2, -3)) == -2, "kadane all-negative wrong")
    println("[Scala Kadane] Maximum subarray sum verified")
  }
}

File: algorithms/13_swift/dp/Kadane.swift
func kadane(_ arr: [Int]) -> Int {
    var best = arr[0], cur = arr[0]
    for x in arr.dropFirst() {
        cur = max(x, cur + x)
        best = max(best, cur)
    }
    return best
}

assert(kadane([-2, 1, -3, 4, -1, 2, 1, -5, 4]) == 6)
assert(kadane([-5, -2, -3]) == -2)
print("[Swift Kadane] Maximum subarray sum verified")
